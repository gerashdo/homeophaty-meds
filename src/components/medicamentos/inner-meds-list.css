
.delete-btn {
    display: none;
}

li:hover .delete-btn {
    display: block;
    transition: all 0.3s ease-in-out;
}