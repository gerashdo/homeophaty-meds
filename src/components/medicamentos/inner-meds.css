


input.inner-input {
    padding: 2px 5px;
    border: 0;
    border-bottom: 1px solid rgb(169, 169, 169);
    width: 100%;
    outline: none;
}

.input.process button {
    border: none;
    background-color: var(--color-primary);
    color: white;
    padding: 0.125rem 0.5rem;
    border-radius: 25%;
    cursor: pointer;
}