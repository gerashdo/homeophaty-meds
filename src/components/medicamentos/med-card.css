.med-card{
    /* border: 1px solid rgb(0, 0, 0); */
    border-radius: 1.25rem;
    box-shadow: 2px 2px 5px rgba(0, 0, 0, 0.1);
    display: flex;
    flex-direction: column;
    height: 20rem;
    gap: 1rem;
    justify-content: space-between;
    min-width: 60%;
    transition: all 0.1s ease;
    overflow: hidden;
}

.med-card:hover {
    box-shadow: 2px 1px 5px rgba(0, 0, 0, 0.2);
    min-width: 62%;
}

.med-card-content {
    display: flex;
    flex-direction: column;
    gap: 1rem;
}


.med-card .med-card-content .title {
    font-weight: var(--font-bold);
    font-size: var(--font-size-subtitle);
}

.med-card .title{
    color: white;
    padding: 1rem;
}

.med-card .title.formula {
    background-color: var(--color-secondary-darker);
}

.med-card .title.normal {
    background-color: var(--color-primary-darker);
}

.med-card-content p{
    color: var(--color-black-lighter);
    padding: 0 2rem;
    text-align: start;

}

.med-card .card-footer{
    padding: 1rem;
}

button a {
    text-decoration: none;
    color: white;
}