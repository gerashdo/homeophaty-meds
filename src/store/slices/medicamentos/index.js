export * from './medicamentoSlice'
export * from './thunks'