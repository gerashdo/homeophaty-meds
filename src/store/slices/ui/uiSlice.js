import { createSlice } from "@reduxjs/toolkit";

const initialState = {
    alertActive: false,
    alertMessage: '',
    alertType: null,
}

export const uiSlice = createSlice({
    name: 'ui',
    initialState: initialState,
    reducers: {
        startAlert: ( state, action ) => {
            state.alertActive = true
            state.alertMessage = action.payload.alertMessage
            state.alertType = action.payload.alertType
        },
        stopAlert: ( state ) => {
            state.alertActive = false
        }
    }
})

export const { 
    startAlert, 
    stopAlert,
} = uiSlice.actions