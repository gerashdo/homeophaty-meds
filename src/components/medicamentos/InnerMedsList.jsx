import React from 'react'
import { CgTrash } from 'react-icons/cg'

import './inner-meds-list.css'

export const InnerMedsList = ({ innerMeds, onRemoveInnerMeds }) => {
  return (
    <ul className='mb-4 flex gap-3'>
      {innerMeds.map(med => (
        <li key={med._id} className='flex border-2 border-gray-300 px-3 py-1 rounded-md  items-center justify-between'>
          <span>{med.name} {med.ch}</span>
          <button className='delete-btn text-gray-400 hover:text-gray-600 p-1' onClick={() => onRemoveInnerMeds(med._id)}><CgTrash /></button>
        </li>
      ))}
    </ul>
  )
}
