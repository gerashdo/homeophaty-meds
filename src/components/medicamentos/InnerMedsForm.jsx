
import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useForm } from '../../hooks/useForm';
import { startAlert } from '../../store/slices/ui';
import { DinamicSelector } from '../iterface/DinamicSelector';

import './inner-meds.css'

// Parametros
// onChangeInnerMeds: Accion a realizar cuando se agregue un nuevo medicamento
// innerMeds: Lista de innerMeds actuales
export const InnerMedsForm = ({ onChangeInnerMeds, innerMeds }) => {
  const dispatch = useDispatch()
  const { medicamentos } = useSelector( state => state.medicamento )
  const [ values, handleChange, setValues, reset ] = useForm({ medicamento: '' })
  const { medicamento } = values
  const [ focused, setFocused ] = useState( false )
  const [ valueId, setValueId ] = useState('')

  const changeInputValue = ( newValue ) => {
    setValues({
      ...values,
      medicamento: newValue
    })
  }

  //TODO: Obtener el objeto del que se dio click y esta en el input
  const handleAddMedicine = (e) => {
    e.preventDefault()

    // Validate that the new innerMed is not in the list already
    // Cuando el medicamento se va a crear los innerMeds tienen id
    // Cuando se va a actualizar un medicamento los innerMeds tienen _id
    if( innerMeds.find( med => med.id === valueId || med._id === valueId ) ){
      return dispatch( startAlert({
        alertMessage: 'El medicamento ya se encuentra en la lista',
        alertType: 'info'
      }) )
    }

    const result = medicamentos.find( med => med.id === valueId )

    if( result ){
      onChangeInnerMeds([
        ...innerMeds,
        { ...result, '_id': result.id }
      ])
    }

    reset()
  }

  return (
    <div 
      className='horizontal-input-button'
    >
      <div className='input process relative'>
        <input
          name='medicamento'
          className='inner-input'
          onChange={ handleChange }
          value={ medicamento }
          onFocus={ () => setFocused( true ) }
          onBlur={ () => setFocused( false ) }
        />
        {
          focused
          &&
          <DinamicSelector 
            resetValues={ medicamentos }
            valueForFilter={ medicamento }
            onSetValue={ changeInputValue }
            onSetValueId={ setValueId }
          />
        }
      </div>
      <button
        className='primary'
        onClick={ handleAddMedicine }
      >
        +
      </button>
    </div>
  )
}
