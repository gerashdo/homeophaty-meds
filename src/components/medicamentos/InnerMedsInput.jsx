import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types'
import { useForm } from '../../hooks/useForm';
import { startAlert } from '../../store/slices/ui';
import { DinamicSelector } from '../iterface/DinamicSelector';
import { searchStringInMed } from '../../helpers';

import './inner-meds.css'

// Parametros
// onChangeInnerMeds: Accion a realizar cuando se agregue un nuevo medicamento
// innerMeds: Lista de innerMeds actuales
export const InnerMedsInput = ({ onChangeInnerMeds, innerMeds = [] }) => {
    const dispatch = useDispatch()
    const { medicamentos } = useSelector( state => state.medicamento )
    const [ values, handleChange, setValues, reset ] = useForm({ medicamento: '' })
    const { medicamento } = values
    const [ focused, setFocused ] = useState( false )

    const changeInputValue = ( newValue ) => {
        setValues({
            ...values,
            medicamento: newValue
        })
    }

    const handleAddMedicine = (e) => {
        e.preventDefault()

        // Si el medicamento ya esta dentro del los inner meds
        if( searchStringInMed( medicamento, innerMeds ).length > 0 ){
            return dispatch( startAlert({
                alertMessage: 'El medicamento ya se encuentra en la lista',
                alertType: 'info'
            }) )
        }

        const medsFound = searchStringInMed( medicamento, medicamentos )

        // Si hay mas de un resultado con el texto introducido en el input
        if( medsFound.length > 1 ){
            return dispatch( startAlert({
                alertMessage: 'Hay m√°s de un medicamento con el nombre que introdujo',
                alertType: 'info'
            }) )
        }
        
        onChangeInnerMeds( medsFound[0] )

        reset()
    }

    return (
        <div 
            className='horizontal-input-button'
        >
        <div className='input process relative'>
            <input
                name='medicamento'
                className='inner-input'
                onChange={ handleChange }
                value={ medicamento }
                onFocus={ () => setFocused( true ) }
                onBlur={ () => setFocused( false ) }
            />
            {
                focused
                &&
                <DinamicSelector 
                    resetValues={ medicamentos }
                    valueForFilter={ medicamento }
                    onSetValue={ changeInputValue }
                />
            }
        </div>
            <button
                className='primary'
                onClick={ handleAddMedicine }
            >
                +
            </button>
        </div>
    )
}

InnerMedsInput.propTypes = {
    onChangeInnerMeds: PropTypes.func.isRequired,
    innerMeds: PropTypes.arrayOf( PropTypes.object )
}