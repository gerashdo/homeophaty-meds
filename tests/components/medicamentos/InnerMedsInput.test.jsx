import { configureStore } from "@reduxjs/toolkit";
import { act, fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import * as uiSlice from "../../../src/store/slices/ui/uiSlice";
import { InnerMedsInput } from "../../../src/components/medicamentos/InnerMedsInput";
import { medicamentoSlice } from "../../../src/store/slices/medicamentos";
import { abrotanum200, abrotanum30, medsListFixture } from "../../fixtures/meds";
import { alertGenericInfo } from "../../fixtures/alertFixtures";

const getMockStore = ( initialState ) => {
    return configureStore({
        reducer: {
            medicamento: medicamentoSlice.reducer,
        },
        preloadedState: {
            medicamento: { ...initialState }
        }
    })
}

// TODO: pendiente
describe('Tests for <InnerMedsInput />', () => {

    beforeEach( () => {
        jest.clearAllMocks()
    })

    afterEach( () => {
        jest.restoreAllMocks()
    })
    
    const onChangeInnerMeds = jest.fn()

    test('should render the component', () => {

        const storeMock = getMockStore({ medicamentos: medsListFixture })

        const { container } = render(
            <InnerMedsInput 
                onChangeInnerMeds={ onChangeInnerMeds }
            />,
            { wrapper: ({ children }) => <Provider store={ storeMock }>{ children }</Provider>}
        )

        expect( container ).toMatchSnapshot()
    });

    test('should render or not depending if the input is focused', () => {

        const storeMock = getMockStore({ medicamentos: medsListFixture })

        const { queryAllByRole } = render(
            <InnerMedsInput 
                onChangeInnerMeds={ onChangeInnerMeds }
            />,
            { wrapper: ({ children }) => <Provider store={ storeMock }>{ children }</Provider>}
        )

        act( () => screen.getByRole('textbox').focus() )
        expect( screen.getAllByRole('listitem' ).length ).toBeGreaterThan( 0 )

        act( () => screen.getByRole('textbox').blur() )
        expect( queryAllByRole( 'listitem' ).length ).toBe( 0 )

    });

    test('should call the onChangeInnerMeds function if the medicine is not in the innerMeds', () => {
        const storeMock = getMockStore({ medicamentos: medsListFixture })

        render(
            <InnerMedsInput 
                onChangeInnerMeds={ onChangeInnerMeds }
            />,
            { wrapper: ({ children }) => <Provider store={ storeMock }>{ children }</Provider>}
        )
        
        const input = screen.getByRole('textbox')
        act( () => input.focus() )
        fireEvent.change( input, { target: { value: `${abrotanum30.name}` }})

        act( () => screen.getByRole('button').click() )
        
        expect( onChangeInnerMeds ).toHaveBeenCalledWith( abrotanum30 )
    });

    test('should render an info alert if the medicine is alredy in the innerMeds', () => {
        const storeMock = getMockStore({ medicamentos: medsListFixture })
        const startAlertMock = jest.spyOn( uiSlice, 'startAlert')

        render(
            <InnerMedsInput 
                onChangeInnerMeds={ onChangeInnerMeds }
                innerMeds={ medsListFixture }
            />,
            { wrapper: ({ children }) => <Provider store={ storeMock }>{ children }</Provider>}
        )
        
        const input = screen.getByRole('textbox')
        act( () => input.focus() )
        fireEvent.change( input, { target: { value: `${abrotanum30.name}` }})

        act( () => screen.getByRole('button').click() )

        expect( onChangeInnerMeds ).not.toHaveBeenCalled()
        expect( startAlertMock ).toHaveBeenCalledWith({ 
            ...alertGenericInfo, 
            alertMessage: 'El medicamento ya se encuentra en la lista'
        })
    });

    test('should render an info alert if there are more than one result from the input text', () => {
        const storeMock = getMockStore({ medicamentos: [ ...medsListFixture, abrotanum200 ] })
        const startAlertMock = jest.spyOn( uiSlice, 'startAlert')

        render(
            <InnerMedsInput 
                onChangeInnerMeds={ onChangeInnerMeds }
            />,
            { wrapper: ({ children }) => <Provider store={ storeMock }>{ children }</Provider>}
        )
        
        const input = screen.getByRole('textbox')
        act( () => input.focus() )
        fireEvent.change( input, { target: { value: `${abrotanum30.name}` }})

        act( () => screen.getByRole('button').click() )

        expect( onChangeInnerMeds ).not.toHaveBeenCalled()
        expect( startAlertMock ).toHaveBeenCalledWith({ 
            ...alertGenericInfo, 
            alertMessage: 'Hay m√°s de un medicamento con el nombre que introdujo'
        })
    }); 
});