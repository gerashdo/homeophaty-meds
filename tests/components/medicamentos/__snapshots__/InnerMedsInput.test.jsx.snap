// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <InnerMedsInput /> should render the component 1`] = `
<div>
  <div
    class="horizontal-input-button"
  >
    <div
      class="input process relative"
    >
      <input
        class="inner-input"
        name="medicamento"
        value=""
      />
    </div>
    <button
      class="primary"
    >
      +
    </button>
  </div>
</div>
`;
